engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala

```diff
@@ -218,41 +219,60 @@ class EngineJobExecutionActor(replyTo: ActorRef,
   }
 
   when(RunningJob) {
-    case Event(hashes: CallCacheHashes, data: SucceededResponseData) =>
+    // Received new hashes with CacheWriteResponseData data, meaning the job succeeded or failed and won't be retried
+    case Event(hashes: CallCacheHashes, data: CacheWriteResponseData) =>
       saveCacheResults(hashes, data)
+    // Received new hashes with ResponsePendingData data, meaning the job is still running
     case Event(hashes: CallCacheHashes, data: ResponsePendingData) =>
       addHashesAndStay(data, hashes)
     case Event(CacheMiss, _) =>
       stay()
     case Event(_: CacheHit, _) =>
       stay()
 
-    case Event(HashError(t), data: SucceededResponseData) =>
+    case Event(HashError(t), data: CacheWriteResponseData) =>
       disableCallCaching(Option(t))
-      saveJobCompletionToJobStore(data.copy(hashes = Option(Failure(t))))
+      saveJobCompletionToJobStore(data.withHashes(hashes = Failure(t)))
     case Event(HashError(t), data: ResponsePendingData) =>
       disableCallCaching(Option(t))
       stay using data.copy(hashes = Option(Failure(t)))
 
+      // Success
+        // All hashes already retrieved - save to the cache
     case Event(response: JobSucceededResponse, data @ ResponsePendingData(_, _, Some(Success(hashes)), _, _)) if effectiveCallCachingMode.writeToCache =>
       eventList ++= response.executionEvents
       saveCacheResults(hashes, data.withSuccessResponse(response))
+        // Some hashes are still missing - waiting for them
     case Event(response: JobSucceededResponse, data @ ResponsePendingData(_, _, None, _, _)) if effectiveCallCachingMode.writeToCache =>
       eventList ++= response.executionEvents
       log.debug(s"Got job result for {}, awaiting hashes", jobTag)
       stay using data.withSuccessResponse(response)
+        // writeToCache is OFF - complete the job
     case Event(response: JobSucceededResponse, data: ResponsePendingData) =>
       eventList ++= response.executionEvents
       saveJobCompletionToJobStore(data.withSuccessResponse(response))
+      
+      // Non-Retryable failure
+        // All hashes already retrieved - save to the cache
+    case Event(response: JobFailedNonRetryableResponse, data @ ResponsePendingData(_, _, Some(Success(hashes)), _, _)) if effectiveCallCachingMode.writeToCache =>
```

Comment:
Before, we didn't need to separate out the cases of a failure coming back with hashes and without hashes.